[dtensor] update public API docs

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127340
* #127339
* #127338

This PR updates the API documentations for the public facing APIs

needs more example for each API but plan to add them in a separate PR

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

wanchaol committed:
[dtensor] update public API docs

This PR updates the API documentations for the public facing APIs

needs more example for each API but plan to add them in a separate PR

[ghstack-poisoned]

awgu commented on torch/distributed/_tensor/api.py:
```diff
@@ -523,8 +523,10 @@ def distribute_tensor(
     """
-    Distribute a torch.Tensor to the `device_mesh` according to the `placements`
-    specified. The rank of `device_mesh` and `placements` must be the same.
+    Distribute a leaf torch.Tensor (i.e. nn.Parameter) to the `device_mesh` according
```
nit: might need to check these renders later to make sure code type is used

wanchaol replied:
I actually haven't add the docs page yet in the pytorch documentation, plan to add a dedicated page in the follow up PRs and check the rendering

awgu commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -37,7 +37,17 @@ def is_partial(self) -> bool:
 class Shard(Placement):
-    # shard placement, shard on a dim
+    """
+    The ``Shard(dim)`` placement describes the DTensor sharding on a specified
```
I think somehow this description should clarify/mention that the `dim` in `Shard(dim)` is the _tensor_ dim across which sharding happens, not the `DeviceMesh` dimension. (I think this might be a common source of confusion.)

wanchaol replied:
yeah good point! updating

wz337 commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -368,9 +384,18 @@ def _replicate_tensor(
     # This is a default _Partial placement with element-wise reduce op
-    # _Partial define three contracts:
-    # 1. _reduce_value: reduce the value of the tensor on the mesh dimension
     # 2. _reduce_shard_value: reduce_scatter the value of the tensor on the mesh dimension
```
nit: #2 -> #1

wz337 commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -368,9 +384,18 @@ def _replicate_tensor(
-    # _Partial define three contracts:
-    # 1. _reduce_value: reduce the value of the tensor on the mesh dimension
     # 2. _reduce_shard_value: reduce_scatter the value of the tensor on the mesh dimension
     # 3. _partition_value: partition the value of a replicated tensor on the mesh dimension
```
ditto

wanchaol committed:
Update on "[dtensor] update public API docs"


This PR updates the API documentations for the public facing APIs

needs more example for each API but plan to add them in a separate PR

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
